docs/proposals/20-ha-control-planes.md

```diff
@@ -0,0 +1,1259 @@
+
+**Active-Passive**
+* Components running in  a `active-passive` mode are expected to have at least **two** replicas, so that there is always one replica on stand-by.
+
+## Gardener Shoot API
+
+### Proposed changes
+
+The following changes to the shoot API are suggested to enable the HA feature for a shoot cluster:
+
+```yaml
+kind: Shoot
+apiVersion: core.gardener.cloud/v1beta1
+spec:
+  controlPlane:
+    highAvailability:
+      enabled: true # by default HA is disabled.
+      failureDomain:
+        type: <single-zone | multi-zone>
+        whenUnsatisfied: <DoNotSchedule | ScheduleAnyway>
+```
+
+An additional capability is provided to the consumer to clearly define the behavior if the chosen type of failure domain cannot be provisioned by gardener due lack of available seeds capable of either hosting a `single-zone` or `multi-zone` shoot control plane.
+
+### Future enhancements
+
+Following additional options influencing scheduling and provisioning of shoot control plane components could be provided in the future:
+
+_Failure tolerance_
+
+In the context of an etcd cluster, a 3 member cluster will only provide a failure tolerance of 1 availability zone (for multi-zone) or node (for single-zone-multi-node). This can still be the default setup, but an option could be provided to make it configurable.
+
+```yaml
+kind: Shoot
+apiVersion: core.gardener.cloud/v1beta1
+spec:
+  controlPlane:
+    highAvailability:
+      enabled: true # by default HA is disabled.
+      type: <single-zone | multi-zone>
+      failureTolerance: <possible values should be in the range 1..2(included)>
+```
+
+## Gardener Scheduler
+
+A scheduling request could be for a shoot with HA or non-HA control plane. It is therefore required to appropriately select a seed. There are two use cases that need to be considered:
+
+**Case #1**
+
+A shoot which has no seed assigned is to be scheduled.
+
+_Proposed Changes_
+
+Introduce a new filtering step in [reconciler](https://github.com/gardener/gardener/blob/4dbba56e24b5999fd728cb84b09a4d6d54f64479/pkg/scheduler/controller/shoot/reconciler.go#L110) which will do the following:
+
+* Segregate available seeds into `multi-zonal` seeds and `non-multi-zonal` seeds.
+* Assign selection priority based on shoot spec HA configuration for the control plane.
+  * Shoot with `multi-zonal` control plane can only be scheduled on `multi-zonal` seed. Therefore for these shoots the target list of seeds will only be the ones which are `multi-zonal`.
+  * Shoot with `single-zonal` control plane can be scheduled either on `single-zonal` seed or `multi-zonal` seeds. In order to find the best fit - first check if there is any available seed in `single-zonal` seeds. If there is one, then that is assigned. If there is no available seed in `non-multi-zonal` seeds, then look for an available seed in `multi-zonal` seeds.
```

Comment:
Really? Even if that single-zone seed is on the other half of the world, but the region where the `single-zone` shoot is scheduled to has "only" free multi-zone seeds? Such an absolute filtering (regardless of the region) seems wrong to me.

Reply:
Hm, I think there is a misunderstanding. The goal was rather that if you have multiple seed candidates in the very last step where the seed with the least allocated shoots are preferred, we thought it'd make sense to also consider that a single-zone seed should be preferred for a single-zone (or non-HA) shoot.

Reply:
Please have a look at the revised section.